// @nervosnetwork/ckb-sdk-rpc@0.109.3/lib/method.js
"use strict";
var _Method_name, _Method_options, _Method_node;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const axios_1 = (0, tslib_1.__importDefault)(require("axios"));
const index_js_1 = require("./exceptions/index.js");
class Method {
    constructor(node, options) {
        _Method_name.set(this, void 0);
        _Method_options.set(this, {
            name: '',
            method: '',
            paramsFormatters: [],
            resultFormatters: undefined,
        });
        _Method_node.set(this, void 0);
        this.call = (...params) => {
            const payload = this.getPayload(...params);
            return (0, axios_1.default)({
                method: 'POST',
                headers: {
                    'content-type': 'application/json',
                },
                data: payload,
                url: (0, tslib_1.__classPrivateFieldGet)(this, _Method_node, "f").url,
                httpAgent: (0, tslib_1.__classPrivateFieldGet)(this, _Method_node, "f").httpAgent,
                httpsAgent: (0, tslib_1.__classPrivateFieldGet)(this, _Method_node, "f").httpsAgent,
            }).then(res => {
                var _a, _b, _c;
                if (res.data.id !== payload.id) {
                    throw new index_js_1.IdNotMatchException(payload.id, res.data.id);
                }
                if (res.data.error) {
                    throw new index_js_1.ResponseException(JSON.stringify(res.data.error));
                }
                return (_c = (_b = (_a = (0, tslib_1.__classPrivateFieldGet)(this, _Method_options, "f")).resultFormatters) === null || _b === void 0 ? void 0 : _b.call(_a, res.data.result)) !== null && _c !== void 0 ? _c : res.data.result;
            });
        };
        this.getPayload = (...params) => {
            const data = params.map((p, i) => ((0, tslib_1.__classPrivateFieldGet)(this, _Method_options, "f").paramsFormatters[i] && (0, tslib_1.__classPrivateFieldGet)(this, _Method_options, "f").paramsFormatters[i](p)) || p);
            const id = Math.round(Math.random() * 10000);
            const payload = {
                id,
                method: (0, tslib_1.__classPrivateFieldGet)(this, _Method_options, "f").method,
                params: data,
                jsonrpc: '2.0',
            };
            return payload;
        };
        (0, tslib_1.__classPrivateFieldSet)(this, _Method_node, node, "f");
        (0, tslib_1.__classPrivateFieldSet)(this, _Method_options, options, "f");
        (0, tslib_1.__classPrivateFieldSet)(this, _Method_name, options.name, "f");
        Object.defineProperty(this.call, 'name', { value: options.name, configurable: false, writable: false });
    }
    get name() {
        return (0, tslib_1.__classPrivateFieldGet)(this, _Method_name, "f");
    }
}
_Method_name = new WeakMap(), _Method_options = new WeakMap(), _Method_node = new WeakMap();
exports.default = Method;
