// @spore-sdk/core@0.2.1/lib/cobuild/action/cluster/transferCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTransferClusterBuildingPacket = exports.generateTransferClusterAction = exports.assembleTransferClusterAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
function assembleTransferClusterAction(clusterInput, clusterOutput) {
    const actions = [];
    const scriptInfos = [];
    const clusterType = clusterOutput.cellOutput.type;
    const clusterTypeHash = lumos_1.utils.computeScriptHash(clusterType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: clusterTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'TransferCluster',
        value: {
            clusterId: clusterType.args,
            from: {
                type: 'Script',
                value: clusterInput.cellOutput.lock,
            },
            to: {
                type: 'Script',
                value: clusterOutput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: clusterTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.assembleTransferClusterAction = assembleTransferClusterAction;
function generateTransferClusterAction(props) {
    let txSkeleton = props.txSkeleton;
    const clusterInput = txSkeleton.get('inputs').get(props.inputIndex);
    const clusterOutput = txSkeleton.get('outputs').get(props.outputIndex);
    return assembleTransferClusterAction(clusterInput, clusterOutput);
}
exports.generateTransferClusterAction = generateTransferClusterAction;
function generateTransferClusterBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateTransferClusterAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: props.useCapacityMarginAsFee ? props.outputIndex : txSkeleton.get('outputs').size - 1,
    });
}
exports.generateTransferClusterBuildingPacket = generateTransferClusterBuildingPacket;
