// @ckb-lumos/lumos@0.22.0-next.5/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TESTNET = exports.MAINNET = void 0;
Object.defineProperty(exports, "createConfig", {
  enumerable: true,
  get: function () {
    return _configManager.createConfig;
  }
});
Object.defineProperty(exports, "createLatestTypeIdResolver", {
  enumerable: true,
  get: function () {
    return _configManager.createLatestTypeIdResolver;
  }
});
Object.defineProperty(exports, "createRpcResolver", {
  enumerable: true,
  get: function () {
    return _configManager.createRpcResolver;
  }
});
Object.defineProperty(exports, "findConfigByScript", {
  enumerable: true,
  get: function () {
    return _configManager.findConfigByScript;
  }
});
Object.defineProperty(exports, "generateGenesisScriptConfigs", {
  enumerable: true,
  get: function () {
    return _configManager.generateGenesisScriptConfigs;
  }
});
Object.defineProperty(exports, "getConfig", {
  enumerable: true,
  get: function () {
    return _configManager.getConfig;
  }
});
Object.defineProperty(exports, "helpers", {
  enumerable: true,
  get: function () {
    return _configManager.helpers;
  }
});
Object.defineProperty(exports, "initializeConfig", {
  enumerable: true,
  get: function () {
    return _configManager.initializeConfig;
  }
});
Object.defineProperty(exports, "nameOfScript", {
  enumerable: true,
  get: function () {
    return _configManager.nameOfScript;
  }
});
Object.defineProperty(exports, "predefined", {
  enumerable: true,
  get: function () {
    return _configManager.predefined;
  }
});
Object.defineProperty(exports, "refreshScriptConfigs", {
  enumerable: true,
  get: function () {
    return _configManager.refreshScriptConfigs;
  }
});
Object.defineProperty(exports, "validateConfig", {
  enumerable: true,
  get: function () {
    return _configManager.validateConfig;
  }
});
var _predefined = require("@ckb-lumos/config-manager/lib/predefined");
var _configManager = require("@ckb-lumos/config-manager");
const TESTNET = _predefined.predefined.AGGRON4;
exports.TESTNET = TESTNET;
const MAINNET = _predefined.predefined.LINA;
exports.MAINNET = MAINNET;
