// @ckb-lumos/lumos@0.22.0-next.5/common-scripts/dao.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CellCollector", {
  enumerable: true,
  get: function () {
    return _dao.CellCollector;
  }
});
Object.defineProperty(exports, "calculateDaoEarliestSince", {
  enumerable: true,
  get: function () {
    return _dao.calculateDaoEarliestSince;
  }
});
Object.defineProperty(exports, "calculateDaoEarliestSinceCompatible", {
  enumerable: true,
  get: function () {
    return _dao.calculateDaoEarliestSinceCompatible;
  }
});
Object.defineProperty(exports, "calculateMaximumWithdraw", {
  enumerable: true,
  get: function () {
    return _dao.calculateMaximumWithdraw;
  }
});
Object.defineProperty(exports, "calculateMaximumWithdrawCompatible", {
  enumerable: true,
  get: function () {
    return _dao.calculateMaximumWithdrawCompatible;
  }
});
Object.defineProperty(exports, "deposit", {
  enumerable: true,
  get: function () {
    return _dao.deposit;
  }
});
Object.defineProperty(exports, "extractDaoDataCompatible", {
  enumerable: true,
  get: function () {
    return _dao.extractDaoDataCompatible;
  }
});
Object.defineProperty(exports, "listDaoCells", {
  enumerable: true,
  get: function () {
    return _dao.listDaoCells;
  }
});
Object.defineProperty(exports, "unlock", {
  enumerable: true,
  get: function () {
    return _dao.unlock;
  }
});
var _dao = require("@ckb-lumos/common-scripts/lib/dao");
