// @ckb-lumos/lumos@0.22.0-next.5/common-scripts/secp256k1-blake160.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CellCollector", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.CellCollector;
  }
});
Object.defineProperty(exports, "injectCapacity", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.injectCapacity;
  }
});
Object.defineProperty(exports, "payFee", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.payFee;
  }
});
Object.defineProperty(exports, "prepareSigningEntries", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.prepareSigningEntries;
  }
});
Object.defineProperty(exports, "setupInputCell", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.setupInputCell;
  }
});
Object.defineProperty(exports, "transfer", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.transfer;
  }
});
Object.defineProperty(exports, "transferCompatible", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake.transferCompatible;
  }
});
var _secp256k1_blake = require("@ckb-lumos/common-scripts/lib/secp256k1_blake160");
