// @ckb-lumos/lumos@0.22.0-next.5/codec/blockchain.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Block", {
  enumerable: true,
  get: function () {
    return _blockchain.Block;
  }
});
Object.defineProperty(exports, "BlockV1", {
  enumerable: true,
  get: function () {
    return _blockchain.BlockV1;
  }
});
Object.defineProperty(exports, "Byte32Vec", {
  enumerable: true,
  get: function () {
    return _blockchain.Byte32Vec;
  }
});
Object.defineProperty(exports, "Bytes", {
  enumerable: true,
  get: function () {
    return _blockchain.Bytes;
  }
});
Object.defineProperty(exports, "BytesOpt", {
  enumerable: true,
  get: function () {
    return _blockchain.BytesOpt;
  }
});
Object.defineProperty(exports, "BytesOptVec", {
  enumerable: true,
  get: function () {
    return _blockchain.BytesOptVec;
  }
});
Object.defineProperty(exports, "BytesVec", {
  enumerable: true,
  get: function () {
    return _blockchain.BytesVec;
  }
});
Object.defineProperty(exports, "CellDep", {
  enumerable: true,
  get: function () {
    return _blockchain.CellDep;
  }
});
Object.defineProperty(exports, "CellDepVec", {
  enumerable: true,
  get: function () {
    return _blockchain.CellDepVec;
  }
});
Object.defineProperty(exports, "CellInput", {
  enumerable: true,
  get: function () {
    return _blockchain.CellInput;
  }
});
Object.defineProperty(exports, "CellInputVec", {
  enumerable: true,
  get: function () {
    return _blockchain.CellInputVec;
  }
});
Object.defineProperty(exports, "CellOutput", {
  enumerable: true,
  get: function () {
    return _blockchain.CellOutput;
  }
});
Object.defineProperty(exports, "CellOutputVec", {
  enumerable: true,
  get: function () {
    return _blockchain.CellOutputVec;
  }
});
Object.defineProperty(exports, "CellbaseWitness", {
  enumerable: true,
  get: function () {
    return _blockchain.CellbaseWitness;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function () {
    return _blockchain.Header;
  }
});
Object.defineProperty(exports, "OutPoint", {
  enumerable: true,
  get: function () {
    return _blockchain.OutPoint;
  }
});
Object.defineProperty(exports, "ProposalShortId", {
  enumerable: true,
  get: function () {
    return _blockchain.ProposalShortId;
  }
});
Object.defineProperty(exports, "ProposalShortIdVec", {
  enumerable: true,
  get: function () {
    return _blockchain.ProposalShortIdVec;
  }
});
Object.defineProperty(exports, "RawHeader", {
  enumerable: true,
  get: function () {
    return _blockchain.RawHeader;
  }
});
Object.defineProperty(exports, "RawTransaction", {
  enumerable: true,
  get: function () {
    return _blockchain.RawTransaction;
  }
});
Object.defineProperty(exports, "Script", {
  enumerable: true,
  get: function () {
    return _blockchain.Script;
  }
});
Object.defineProperty(exports, "ScriptOpt", {
  enumerable: true,
  get: function () {
    return _blockchain.ScriptOpt;
  }
});
Object.defineProperty(exports, "Transaction", {
  enumerable: true,
  get: function () {
    return _blockchain.Transaction;
  }
});
Object.defineProperty(exports, "TransactionVec", {
  enumerable: true,
  get: function () {
    return _blockchain.TransactionVec;
  }
});
Object.defineProperty(exports, "Uint128", {
  enumerable: true,
  get: function () {
    return _number.Uint128;
  }
});
Object.defineProperty(exports, "Uint256", {
  enumerable: true,
  get: function () {
    return _number.Uint256;
  }
});
Object.defineProperty(exports, "Uint32", {
  enumerable: true,
  get: function () {
    return _number.Uint32;
  }
});
Object.defineProperty(exports, "Uint64", {
  enumerable: true,
  get: function () {
    return _number.Uint64;
  }
});
Object.defineProperty(exports, "UncleBlock", {
  enumerable: true,
  get: function () {
    return _blockchain.UncleBlock;
  }
});
Object.defineProperty(exports, "UncleBlockVec", {
  enumerable: true,
  get: function () {
    return _blockchain.UncleBlockVec;
  }
});
Object.defineProperty(exports, "WitnessArgs", {
  enumerable: true,
  get: function () {
    return _blockchain.WitnessArgs;
  }
});
var _number = require("@ckb-lumos/codec/lib/number");
var _blockchain = require("@ckb-lumos/base/lib/blockchain");
