// @ckb-lumos/codec@0.23.0/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createBytesCodec: true,
  createFixedBytesCodec: true,
  isFixedCodec: true,
  bytes: true,
  number: true,
  molecule: true
};
exports.bytes = void 0;
Object.defineProperty(exports, "createBytesCodec", {
  enumerable: true,
  get: function () {
    return _base.createBytesCodec;
  }
});
Object.defineProperty(exports, "createFixedBytesCodec", {
  enumerable: true,
  get: function () {
    return _base.createFixedBytesCodec;
  }
});
Object.defineProperty(exports, "isFixedCodec", {
  enumerable: true,
  get: function () {
    return _base.isFixedCodec;
  }
});
exports.number = exports.molecule = void 0;
var _base = require("./base");
var _highOrder = require("./high-order");
Object.keys(_highOrder).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _highOrder[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _highOrder[key];
    }
  });
});
var _bytes = _interopRequireWildcard(require("./bytes"));
exports.bytes = _bytes;
var _number = _interopRequireWildcard(require("./number"));
exports.number = _number;
var _molecule = _interopRequireWildcard(require("./molecule"));
exports.molecule = _molecule;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
