// @spore-sdk/core@0.2.1/lib/cobuild/action/clusterAgent/transferClusterAgent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTransferClusterAgentBuildingPacket = exports.generateTransferClusterAgentAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
function generateTransferClusterAgentAction(props) {
    const actions = [];
    const scriptInfos = [];
    let txSkeleton = props.txSkeleton;
    const clusterAgentInput = txSkeleton.get('inputs').get(props.inputIndex);
    const clusterAgentOutput = txSkeleton.get('outputs').get(props.outputIndex);
    const clusterAgentType = clusterAgentOutput.cellOutput.type;
    const clusterAgentTypeHash = lumos_1.utils.computeScriptHash(clusterAgentType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: clusterAgentTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'TransferClusterAgent',
        value: {
            clusterId: clusterAgentType.args.slice(0, 66),
            from: {
                type: 'Script',
                value: clusterAgentInput.cellOutput.lock,
            },
            to: {
                type: 'Script',
                value: clusterAgentOutput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: clusterAgentTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.generateTransferClusterAgentAction = generateTransferClusterAgentAction;
function generateTransferClusterAgentBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateTransferClusterAgentAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: props.useCapacityMarginAsFee ? props.outputIndex : txSkeleton.get('outputs').size - 1,
    });
}
exports.generateTransferClusterAgentBuildingPacket = generateTransferClusterAgentBuildingPacket;
