// @ckb-lumos/lumos/common-scripts/secp256k1-blake160-multisig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CellCollector", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.CellCollector;
  }
});
Object.defineProperty(exports, "injectCapacity", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.injectCapacity;
  }
});
Object.defineProperty(exports, "payFee", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.payFee;
  }
});
Object.defineProperty(exports, "prepareSigningEntries", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.prepareSigningEntries;
  }
});
Object.defineProperty(exports, "serializeMultisigScript", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.serializeMultisigScript;
  }
});
Object.defineProperty(exports, "setupInputCell", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.setupInputCell;
  }
});
Object.defineProperty(exports, "transfer", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.transfer;
  }
});
Object.defineProperty(exports, "transferCompatible", {
  enumerable: true,
  get: function () {
    return _secp256k1_blake160_multisig.transferCompatible;
  }
});
var _secp256k1_blake160_multisig = require("@ckb-lumos/common-scripts/lib/secp256k1_blake160_multisig");
