// @ckb-lumos/rpc@0.22.0-next.5/lib/types/rpc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RPC = void 0;
/**
 * @see https://github.com/nervosnetwork/ckb/blob/develop/util/jsonrpc-types/src/blockchain.rs
 */
/* eslint-disable  @typescript-eslint/no-namespace */
let RPC;
exports.RPC = RPC;
(function (_RPC) {
  // this is a type to mapping the `HashMap`, `BTreeMap` in `jsonrpc-types`
  // there are some returns of CKB RPC are in this format, like `Softfork`
  let TransactionStatus = /*#__PURE__*/function (TransactionStatus) {
    TransactionStatus["Pending"] = "pending";
    TransactionStatus["Proposed"] = "proposed";
    TransactionStatus["Committed"] = "committed";
    TransactionStatus["Unknown"] = "unknown";
    TransactionStatus["Rejected"] = "rejected";
    return TransactionStatus;
  }({});
  _RPC.TransactionStatus = TransactionStatus;
  /**
   * @deprecated please migrate to {@link ChainInfo}
   */
  /* https://github.com/nervosnetwork/ckb/blob/develop/util/jsonrpc-types/src/info.rs */
})(RPC || (exports.RPC = RPC = {}));
