// @spore-sdk/core@0.2.1/lib/cobuild/action/clusterProxy/createClusterProxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCreateClusterProxyBuildingPacket = exports.generateCreateClusterProxyAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
const referenceCluster_1 = require("../cluster/referenceCluster");
const codec_2 = require("../../../codec");
function generateCreateClusterProxyAction(props) {
    const actions = [];
    const scriptInfos = [];
    let txSkeleton = props.txSkeleton;
    const clusterProxyOutput = txSkeleton.get('outputs').get(props.outputIndex);
    const clusterProxyType = clusterProxyOutput.cellOutput.type;
    const clusterProxyTypeHash = lumos_1.utils.computeScriptHash(clusterProxyType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: clusterProxyTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'CreateClusterProxy',
        value: {
            clusterId: clusterProxyOutput.data,
            clusterProxyId: (0, codec_2.unpackToRawClusterProxyArgs)(clusterProxyType.args).id,
            to: {
                type: 'Script',
                value: clusterProxyOutput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: clusterProxyTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    const clusterAction = (0, referenceCluster_1.generateReferenceClusterAction)({
        txSkeleton,
        referenceType: props.reference.referenceType,
        cluster: props.reference.cluster,
    });
    actions.push(...clusterAction.actions);
    scriptInfos.push(...clusterAction.scriptInfos);
    return {
        actions,
        scriptInfos,
    };
}
exports.generateCreateClusterProxyAction = generateCreateClusterProxyAction;
function generateCreateClusterProxyBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateCreateClusterProxyAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: txSkeleton.get('outputs').size - 1,
    });
}
exports.generateCreateClusterProxyBuildingPacket = generateCreateClusterProxyBuildingPacket;
