// @spore-sdk/core@0.2.1/lib/cobuild/action/clusterProxy/meltClusterProxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMeltClusterProxyBuildingPacket = exports.generateMeltClusterProxyAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
const codec_2 = require("../../../codec");
function generateMeltClusterProxyAction(props) {
    const actions = [];
    const scriptInfos = [];
    let txSkeleton = props.txSkeleton;
    const clusterProxyInput = txSkeleton.get('inputs').get(props.inputIndex);
    const clusterProxyType = clusterProxyInput.cellOutput.type;
    const clusterProxyTypeHash = lumos_1.utils.computeScriptHash(clusterProxyType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: clusterProxyTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'MeltClusterProxy',
        value: {
            clusterId: clusterProxyInput.data,
            clusterProxyId: (0, codec_2.unpackToRawClusterProxyArgs)(clusterProxyType.args).id,
            from: {
                type: 'Script',
                value: clusterProxyInput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: clusterProxyTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.generateMeltClusterProxyAction = generateMeltClusterProxyAction;
function generateMeltClusterProxyBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateMeltClusterProxyAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: txSkeleton.get('outputs').size - 1,
    });
}
exports.generateMeltClusterProxyBuildingPacket = generateMeltClusterProxyBuildingPacket;
