// @spore-sdk/core/lib/api/joints/clusterAgent/injectLiveClusterAgentCell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectLiveClusterAgentCell = void 0;
const helper_1 = require("@ckb-lumos/common-scripts/lib/helper");
const config_1 = require("../../../config");
const helpers_1 = require("../../../helpers");
async function injectLiveClusterAgentCell(props) {
    // Env
    const config = props.config ?? (0, config_1.getSporeConfig)();
    const clusterAgentCell = props.cell;
    // TransactionSkeleton
    let txSkeleton = props.txSkeleton;
    // Check the target cell's type
    const cellType = clusterAgentCell.cellOutput.type;
    const clusterAgentScript = (0, config_1.getSporeScript)(config, 'ClusterAgent', cellType);
    if (!cellType || !clusterAgentScript) {
        throw new Error('Cannot inject ClusterAgent because target cell is not a supported version of ClusterAgent');
    }
    // Add the target cell to Transaction.inputs (and outputs if needed)
    const setupCellResult = await (0, helpers_1.setupCell)({
        txSkeleton,
        input: clusterAgentCell,
        config: config.lumos,
        addOutput: props.addOutput,
        updateOutput(cell) {
            if (props.capacityMargin !== void 0) {
                cell = (0, helpers_1.setAbsoluteCapacityMargin)(cell, props.capacityMargin);
            }
            if (props.updateOutput instanceof Function) {
                cell = props.updateOutput(cell);
            }
            return cell;
        },
        defaultWitness: props.defaultWitness,
        updateWitness: props.updateWitness,
        since: props.since,
    });
    txSkeleton = setupCellResult.txSkeleton;
    // If the target cell has been added to Transaction.outputs
    if (props.addOutput) {
        // Make sure the cell's output has declared enough capacity
        const output = txSkeleton.get('outputs').get(setupCellResult.outputIndex);
        (0, helpers_1.assetCellMinimalCapacity)(output);
        // Fix the cell's output index
        txSkeleton = txSkeleton.update('fixedEntries', (fixedEntries) => {
            return fixedEntries.push({
                field: 'outputs',
                index: setupCellResult.outputIndex,
            });
        });
    }
    // Add ClusterAgent required cellDeps
    txSkeleton = (0, helper_1.addCellDep)(txSkeleton, clusterAgentScript.cellDep);
    return {
        txSkeleton,
        inputIndex: setupCellResult.inputIndex,
        outputIndex: setupCellResult.outputIndex,
    };
}
exports.injectLiveClusterAgentCell = injectLiveClusterAgentCell;
