// @spore-sdk/core@0.2.1/lib/cobuild/action/spore/createSpore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCreateSporeBuildingPacket = exports.generateCreateSporeAction = exports.assembleCreateSporeAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
const referenceCluster_1 = require("../cluster/referenceCluster");
const referenceClusterAgent_1 = require("../clusterAgent/referenceClusterAgent");
function assembleCreateSporeAction(sporeOutput) {
    const actions = [];
    const scriptInfos = [];
    const sporeType = sporeOutput.cellOutput.type;
    const sporeTypeHash = lumos_1.utils.computeScriptHash(sporeType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: sporeTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'CreateSpore',
        value: {
            sporeId: sporeType.args,
            dataHash: lumos_1.utils.ckbHash(sporeOutput.data),
            to: {
                type: 'Script',
                value: sporeOutput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: sporeTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.assembleCreateSporeAction = assembleCreateSporeAction;
function generateCreateSporeAction(props) {
    let txSkeleton = props.txSkeleton;
    const sporeOutput = txSkeleton.get('outputs').get(props.outputIndex);
    let { actions, scriptInfos } = assembleCreateSporeAction(sporeOutput);
    if (props.reference.referenceTarget === 'clusterAgent') {
        const clusterAction = (0, referenceClusterAgent_1.generateReferenceClusterAgentAction)({
            txSkeleton,
            referenceType: props.reference.referenceType,
            clusterAgent: props.reference.clusterAgent,
        });
        actions.push(...clusterAction.actions);
        scriptInfos.push(...clusterAction.scriptInfos);
    }
    if (props.reference.referenceTarget === 'cluster') {
        const clusterAction = (0, referenceCluster_1.generateReferenceClusterAction)({
            txSkeleton,
            referenceType: props.reference.referenceType,
            cluster: props.reference.cluster,
        });
        actions.push(...clusterAction.actions);
        scriptInfos.push(...clusterAction.scriptInfos);
    }
    return {
        actions,
        scriptInfos,
    };
}
exports.generateCreateSporeAction = generateCreateSporeAction;
function generateCreateSporeBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateCreateSporeAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: txSkeleton.get('outputs').size - 1,
    });
}
exports.generateCreateSporeBuildingPacket = generateCreateSporeBuildingPacket;
