// @spore-sdk/core@0.2.1/lib/cobuild/action/spore/meltSpore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMeltSporeBuildingPacket = exports.generateMeltSporeAction = exports.assembleMeltSporeAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
function assembleMeltSporeAction(sporeInput) {
    const actions = [];
    const scriptInfos = [];
    const sporeType = sporeInput.cellOutput.type;
    const sporeTypeHash = lumos_1.utils.computeScriptHash(sporeType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: sporeTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'MeltSpore',
        value: {
            sporeId: sporeType.args,
            from: {
                type: 'Script',
                value: sporeInput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: sporeTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.assembleMeltSporeAction = assembleMeltSporeAction;
function generateMeltSporeAction(props) {
    let txSkeleton = props.txSkeleton;
    const sporeInput = txSkeleton.get('inputs').get(props.inputIndex);
    return assembleMeltSporeAction(sporeInput);
}
exports.generateMeltSporeAction = generateMeltSporeAction;
function generateMeltSporeBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateMeltSporeAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: txSkeleton.get('outputs').size - 1,
    });
}
exports.generateMeltSporeBuildingPacket = generateMeltSporeBuildingPacket;
