// @spore-sdk/core@0.2.1/lib/api/joints/clusterProxy/injectLiveClusterProxyCell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectLiveClusterProxyCell = void 0;
const helper_1 = require("@ckb-lumos/common-scripts/lib/helper");
const config_1 = require("../../../config");
const helpers_1 = require("../../../helpers");
async function injectLiveClusterProxyCell(props) {
    // Env
    const config = props.config ?? (0, config_1.getSporeConfig)();
    const clusterProxyCell = props.cell;
    // TransactionSkeleton
    let txSkeleton = props.txSkeleton;
    // Check target cell's type
    const cellType = clusterProxyCell.cellOutput.type;
    const clusterProxyScript = (0, config_1.getSporeScript)(config, 'ClusterProxy', cellType);
    if (!cellType || !clusterProxyScript) {
        throw new Error('Cannot inject ClusterProxy because target cell is not a supported version of ClusterProxy');
    }
    // Add target cell to Transaction.inputs (and outputs if needed)
    const setupCellResult = await (0, helpers_1.setupCell)({
        txSkeleton,
        input: clusterProxyCell,
        addOutput: props.addOutput,
        updateOutput(cell) {
            if (props.capacityMargin !== void 0) {
                cell = (0, helpers_1.setAbsoluteCapacityMargin)(cell, props.capacityMargin);
            }
            if (props.updateOutput instanceof Function) {
                cell = props.updateOutput(cell);
            }
            return cell;
        },
        defaultWitness: props.defaultWitness,
        updateWitness: props.updateWitness,
        since: props.since,
        config: config.lumos,
    });
    txSkeleton = setupCellResult.txSkeleton;
    // If the target cell is added to Transaction.outputs
    if (props.addOutput) {
        // Make sure the cell's output has declared enough capacity
        const output = txSkeleton.get('outputs').get(setupCellResult.outputIndex);
        (0, helpers_1.assetCellMinimalCapacity)(output);
        // Fix the cell's output index
        txSkeleton = txSkeleton.update('fixedEntries', (fixedEntries) => {
            return fixedEntries.push({
                field: 'outputs',
                index: setupCellResult.outputIndex,
            });
        });
    }
    // Add ClusterProxy required cellDeps
    txSkeleton = (0, helper_1.addCellDep)(txSkeleton, clusterProxyScript.cellDep);
    return {
        txSkeleton,
        inputIndex: setupCellResult.inputIndex,
        outputIndex: setupCellResult.outputIndex,
    };
}
exports.injectLiveClusterProxyCell = injectLiveClusterProxyCell;
