// @spore-sdk/core@0.2.1/lib/codec/spore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unpackToRawSporeData = exports.packRawSporeData = exports.SporeData = void 0;
const base_1 = require("@ckb-lumos/base");
const codec_1 = require("@ckb-lumos/codec");
const helpers_1 = require("../helpers");
exports.SporeData = codec_1.molecule.table({
    contentType: base_1.blockchain.Bytes,
    content: base_1.blockchain.Bytes,
    clusterId: base_1.blockchain.BytesOpt,
}, ['contentType', 'content', 'clusterId']);
function packRawSporeData(packable) {
    return exports.SporeData.pack({
        contentType: (0, helpers_1.bytifyRawString)(packable.contentType),
        content: packable.content,
        clusterId: packable.clusterId,
    });
}
exports.packRawSporeData = packRawSporeData;
function unpackToRawSporeData(unpackable) {
    const unpacked = exports.SporeData.unpack(unpackable);
    return {
        contentType: (0, helpers_1.bufferToRawString)(unpacked.contentType),
        content: unpacked.content,
        clusterId: unpacked.clusterId,
    };
}
exports.unpackToRawSporeData = unpackToRawSporeData;
