// @spore-sdk/core/lib/cobuild/action/cluster/createCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCreateClusterBuildingPacket = exports.generateCreateClusterAction = exports.assembleCreateClusterAction = void 0;
const lumos_1 = require("@ckb-lumos/lumos");
const codec_1 = require("@ckb-lumos/codec");
const sporeAction_1 = require("../../codec/sporeAction");
const buildingPacket_1 = require("../../codec/buildingPacket");
const buildingPacket_2 = require("../../base/buildingPacket");
const sporeScriptInfo_1 = require("../../base/sporeScriptInfo");
function assembleCreateClusterAction(clusterOutput) {
    const actions = [];
    const scriptInfos = [];
    const clusterType = clusterOutput.cellOutput.type;
    const clusterTypeHash = lumos_1.utils.computeScriptHash(clusterType);
    const scriptInfo = (0, sporeScriptInfo_1.createSporeScriptInfoFromTemplate)({
        scriptHash: clusterTypeHash,
    });
    scriptInfos.push(scriptInfo);
    const actionData = sporeAction_1.SporeAction.pack({
        type: 'CreateCluster',
        value: {
            clusterId: clusterType.args,
            dataHash: lumos_1.utils.ckbHash(clusterOutput.data),
            to: {
                type: 'Script',
                value: clusterOutput.cellOutput.lock,
            },
        },
    });
    actions.push({
        scriptInfoHash: lumos_1.utils.ckbHash(buildingPacket_1.ScriptInfo.pack(scriptInfo)),
        scriptHash: clusterTypeHash,
        data: codec_1.bytes.hexify(actionData),
    });
    return {
        actions,
        scriptInfos,
    };
}
exports.assembleCreateClusterAction = assembleCreateClusterAction;
function generateCreateClusterAction(props) {
    let txSkeleton = props.txSkeleton;
    const clusterOutput = txSkeleton.get('outputs').get(props.outputIndex);
    return assembleCreateClusterAction(clusterOutput);
}
exports.generateCreateClusterAction = generateCreateClusterAction;
function generateCreateClusterBuildingPacket(props) {
    let txSkeleton = props.txSkeleton;
    const action = generateCreateClusterAction(props);
    return (0, buildingPacket_2.createRawBuildingPacket)({
        txSkeleton,
        actions: action.actions,
        scriptInfos: action.scriptInfos,
        changeOutput: txSkeleton.get('outputs').size - 1,
    });
}
exports.generateCreateClusterBuildingPacket = generateCreateClusterBuildingPacket;
