// @ckb-lumos/lumos@0.23.0/common-scripts/anyone-can-pay.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CellCollector", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.CellCollector;
  }
});
Object.defineProperty(exports, "injectCapacity", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.injectCapacity;
  }
});
Object.defineProperty(exports, "prepareSigningEntries", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.prepareSigningEntries;
  }
});
Object.defineProperty(exports, "setupInputCell", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.setupInputCell;
  }
});
Object.defineProperty(exports, "setupOutputCell", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.setupOutputCell;
  }
});
Object.defineProperty(exports, "withdraw", {
  enumerable: true,
  get: function () {
    return _anyone_can_pay.withdraw;
  }
});
var _anyone_can_pay = require("@ckb-lumos/common-scripts/lib/anyone_can_pay");
