// @ckb-lumos/lumos@0.23.0/common-scripts/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "injectCapacity", {
  enumerable: true,
  get: function () {
    return _common.injectCapacity;
  }
});
Object.defineProperty(exports, "payFee", {
  enumerable: true,
  get: function () {
    return _common.payFee;
  }
});
Object.defineProperty(exports, "payFeeByFeeRate", {
  enumerable: true,
  get: function () {
    return _common.payFeeByFeeRate;
  }
});
Object.defineProperty(exports, "prepareSigningEntries", {
  enumerable: true,
  get: function () {
    return _common.prepareSigningEntries;
  }
});
Object.defineProperty(exports, "registerCustomLockScriptInfos", {
  enumerable: true,
  get: function () {
    return _common.registerCustomLockScriptInfos;
  }
});
Object.defineProperty(exports, "setupInputCell", {
  enumerable: true,
  get: function () {
    return _common.setupInputCell;
  }
});
Object.defineProperty(exports, "transfer", {
  enumerable: true,
  get: function () {
    return _common.transfer;
  }
});
var _common = require("@ckb-lumos/common-scripts/lib/common");
